Library/OcCpuLib/FrequencyDetect.c

```diff
@@ -225,6 +228,57 @@ InternalCalculateTSCFromPMTimer (
   return TSCFrequency;
 }
 
+UINT64
+InternalCalculateTSCFromApplePlatformInfo (
+  OUT  UINT64  *FSBFrequency  OPTIONAL
+  )
+{
+  EFI_STATUS                             Status;
+  APPLE_PLATFORM_INFO_DATABASE_PROTOCOL  *PlatformInfo;
+  UINT32                                 Size;
+  UINT64                                 FsbFreq;
+
+  Status = gBS->LocateProtocol (
+    &gApplePlatformInfoDatabaseProtocolGuid,
+    NULL,
+    (VOID **) &PlatformInfo
+    );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCCPU: Failed to locate ApplePlatformInfo protocol - %r\n", Status));
+    return 0;
+  }
+
+  Status = PlatformInfo->GetFirstDataSize (
+    PlatformInfo,
+    &gAppleFsbFrequencyPlatformInfoGuid,
+    &Size
+    );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCCPU: Failed to get first data size - %r\n", Status));
+    return 0;
+  }
+
+  Status = PlatformInfo->GetFirstData (
+    PlatformInfo,
+    &gAppleFsbFrequencyPlatformInfoIndexHobGuid,
+    &FsbFreq,
```

Comment:
Before reading to `FsbFreq` please initialise it with 0. As if we have `UINT32` (4 byte) FsbFrequency, then higher bits will not be set to initialised values.